### PR TITLE
Improve rustdoc const bounds

### PR DESCRIPTION
**Pull request by fee1-dead**
- Rustdoc no longer displays `~const` in trait bounds, because it currently means nothing for stable users, and because we still haven't decided on the final syntax yet.
 - Rustdoc will hide trait bounds where the trait is `Drop` AND it is `~const`, i.e. `~const Drop` bounds because it has no effect on stable users as well.
 - Because of additional logic that hides the whole `where` statement where it consists of `~const Drop` bounds (so it doesn't display `struct Foo<T>() where ;` like that), bounds that have no trait e.g. `where [T; N+1]: ;` are also hidden.

Cherry-picked from #92433.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Looks good to me overall, thanks! Your PR seems to done a bit more than what the test you updated seems to check. Can you add more checks please?

Also cc @camelid in case you want to take a second look. :)

File: src/librustdoc/html/format.rs [hunk 1]
```diff
@@ -249,10 +252,80 @@ crate fn print_where_clause<'a, 'tcx: 'a>(
...
+                                format!(
+                                    "for&lt;{:#}&gt; ",
+                                    comma_sep(bound_params.iter().map(|lt| lt.print()), true)
+                                )
+                            }
+                            _ => format!(
+                                "for&lt;{}&gt; ",
+                                comma_sep(bound_params.iter().map(|lt| lt.print()), true)
+                            ),
+                        };
```
Comment (camelid) [inline src/librustdoc/html/format.rs:280]: I think this would be clearer as an `if`-`else` with `is_empty()`.

File: src/librustdoc/html/format.rs [hunk 2]
```diff
@@ -262,72 +335,11 @@ crate fn print_where_clause<'a, 'tcx: 'a>(
...
-                    } else {
-                        clause.push_str(&format!("{} == {}", lhs.print(cx), rhs.print(cx),));
-                    }
-                }
-            }
-
-            if i < gens.where_predicates.len() - 1 || end_newline {
-                clause.push(',');
-            }
-        }
```
Comment (camelid) [inline src/librustdoc/html/format.rs:328]: Why did you move all this code to the beginning of the function?

**Review by camelid — COMMENTED** [review]
Please also add some tests of the expected behavior, e.g. for skipping the `~const` part or excluding `~const Drop`.

File: src/librustdoc/html/format.rs (hunk 2 above)
Comment (fee1-dead) [reply src/librustdoc/html/format.rs:345]: Because I needed to know if there are predicates at all (some could be filtered out) before continuing the formatting.

**Comment by fee1-dead** [comment]
I have addressed the comments and added a test for the display.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks! Waiting for @camelid now. ;)

**Comment by camelid** [comment]
Sorry, I don't have time to review the code. Could you summarize the current version of the changes in this PR?

**Comment by fee1-dead** [comment]
- Hide ~const in rustdoc, it is provisional syntax.
- Hide ~const Drop bounds completely because they do not mean anything to stable users.

**Comment by fee1-dead** [comment]
It also hides const generic wf bounds: [T; SIZE]: (no trait) will be hidden

**Comment by camelid** [comment]
Sounds reasonable. Can you add something to the tracking issue for `~const` about reconsidering the rustdoc display when stabilizing? One more thing: are `~const Trait` bounds no-ops for stable users or are they interpreted as regular `Trait` bounds?

**Comment by fee1-dead** [comment]
Sure, I'll update the issue when I have access to a computer.

As for : ~const Trait they behave as : Trait to stable users.

**Comment by camelid** [comment]
The changes sound good to me. I haven't reviewed the code, but as long as Guillaume has, I don't think I need to. It'd be good to update the PR description with your extended description of the changes, but other than that I think this should be good.

**Comment by GuillaumeGomez** [comment]
Please ping me once the PR description has been updated so I can approve it.

**Comment by fee1-dead** [comment]
@GuillaumeGomez: I have updated the PR description.

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r=

**Comment by GuillaumeGomez** [comment]
Second time will work better...

@bors r+
